syntax = "proto3";

package inworld.ndkData;
option csharp_namespace = "Inworld.NDK";

import "google/protobuf/any.proto";
import "google/protobuf/timestamp.proto";

message Token {
  string token = 1;
  string type = 2;
  google.protobuf.Timestamp expirationTime = 3;
  string sessionId = 4;

  bool IsValid = 5;
}

message LoadSceneRequest {
  ClientRequest client = 1;
  UserRequest user = 2;
  CapabilitiesRequest capabilities = 3;
  UserSettings userSetting = 4;
}

message LoadSceneResponse {
  repeated InworldCharacterData agents = 1;
  string key = 2;
  google.protobuf.Any previousState = 3; // TODO: Solve packets from saved data.
}

message UserRequest {
  string name = 1;
}

message UserSettings {
  bool viewTranscriptConsent = 1;
  PlayerProfile playerProfile = 2;
}

message PlayerProfile {
  repeated Fields fields = 1;
}

message Fields {
  string fieldId = 1;
  string fieldValue = 2;
}

message ClientRequest {
  string id = 1;
  string version = 2;
}

message CapabilitiesRequest {
  bool audio = 1;
  bool emotions = 2;
  bool interruptions = 3;
  bool narratedActions = 4;
  bool silence = 5;
  bool text = 6;
  bool triggers = 7;
  bool continuation = 8;
  bool turnBasedStt = 9;
  bool phonemeInfo = 10;
  bool animations = 11;
  bool gestures = 12;
  bool emotionstreaming = 13;
  bool silenceevents = 14;
  bool loadsceneinsession = 15;
}

message InworldCharacterData {
  string agentId = 1;
  string brainName = 2;
  string givenName = 3;
  CharacterAssets characterAssets = 4;
  bytes thumbnail = 5;
}

message CharacterAssets {
  string rpmModelUri = 1;
  string rpmImageUriPortrait = 2;
  string rpmImageUriPosture = 3;
  string avatarImg = 4;
  string avatarImgOriginal = 5;
}

message ClientOptions {
    string server_url = 2;
    string scene_name = 3;
    string api_key = 4;
    string api_secret = 5;
    string player_name = 6;
    string token = 7;
    string session_id = 8;
    string user_id = 9;
    CapabilitiesRequest capabilities = 10;
}

message AgentInfo {
    string BrainName = 1;
    string AgentId = 2;
    string GivenName = 3;
}

message SessionInfo {
    string SessionId = 1;
    string Token = 2;
    int64 ExpirationTime = 3;
    bool IsValid = 4;
}

message AgentInfoArray {
    repeated AgentInfo agent_info_list = 1;
}

enum ConnectionState {
    Idle = 0;
    Connecting = 1;
    Connected = 2;
    Failed = 3;
    Paused = 4;
    Disconnected = 5;
    Reconnecting = 6;
}
